### PR TITLE
feat: store items in flat storage deltas separately

### PR DESCRIPTION
**Pull request by Longarithm**
Change format for `FlatStorageDelta` as we discussed in https://github.com/near/nearcore/issues/8436 and offline, part of https://github.com/near/nearcore/issues/8436.

Now, instead of storing the whole hashmap for a (shard id, block hash) pair, we store each key-value pair separately, using encoded `FlatStateDeltaKey`. Because the current code assumes that flat state delta always exists, we also store an empty key-value pair indicating the delta existence. [Experiments](https://near.zulipchat.com/#narrow/stream/345766-pagoda.2Fstorage.2Fflat-storage/topic/limiting.20RAM.20usage/near/326107602) already showed that this helps to get rid of growing mem usage at least, just slowing down the node. Even if we change the format, it still makes more sense to store keys separately and avoid reading the whole delta each time.

With that change, it also is more convenient to move methods like `get_delta` from `store_helper` inside `FlatStateDelta` so we could change delta format easier if we come up with more optimizations.

## Testing

* Simple `test_encode_decode` test for delta format
* Existing flat storage tests, e.g. `flat_storage_state_sanity`

### CODE REVIEW

**Review by mzhangmzz — COMMENTED** [review]
I'm confused how this PR limits the memory usage of deltas. It only changes the storage format of how deltas are stored, but doesn't actually limit the deltas in memory. Do you plan to change that?

In this PR, I see that even though deltas are stored in storage in key value pairs separately, but deltas are still loaded into memory as a whole in `FlatStateDelta::read_from_store` function. I want to confirm that you don't plan to change that because I just realized a potential issue with reading delta key value pairs from storage separately instead of loading it entirely.  If we have a in memory cache for deltas in memory and only load/purge (block_hash, key) keys into this cache on demand in the `get_ref` function, then there will be a cache miss every time if a key is not in a block delta. In the worse case, even if a block has no change and has an empty delta, all lookups for the key in this block will not be found in the cache and must hit storage. If we always load deltas into memory as a whole and also stores information of which block deltas are in memory, we can avoid that.

**Review by mzhangmzz — COMMENTED** [review]
LGTM!

**Review by mzhangmzz — APPROVED** [review]

File: core/store/src/flat_state.rs [hunk 1]
```diff
@@ -331,18 +331,56 @@ mod imp {
...
+    pub key: Vec<u8>,
+}
+
+impl FlatStateDeltaKey {
+    #[allow(unused)]
+    /// Should return prefix of all `FlatStateDeltaKey`s corresponding to one flat storage delta.
+    fn db_prefix(shard_id: ShardId, block_hash: CryptoHash) -> Vec<u8> {
+        let mut out = Vec::new();
+        out.extend(shard_id.to_be_bytes());
+        out.extend(block_hash.as_bytes());
```
Comment (pugachAG) [inline core/store/src/flat_state.rs:361]: this is part of encoding logic which is duplicated in `encode`, lets extract this to a separate method, something like `fn append_prefix(out: &mut Vec<u8>, shard_id: ShardId, block_hash: CryptoHash)` or `fn encode_prefix(shard_id: ShardId, block_hash: CryptoHash) -> [u8;40]` and then append it

**Review by pugachAG — CHANGES_REQUESTED** [review]
> we also store an empty key-value pair indicating the delta existence

This looks like a hack to me. Can we instead just read prefix until we reach the next block?

**Comment by Longarithm** [comment]
> Can we instead just read prefix until we reach the next block?

Could you clarify what do you mean?

Without that, there is no way to distinguish between cases "delta was saved but it is empty" and "delta was not saved yet". I think there is a small benefit to have it, because if delta was not saved incorrectly, we can throw an error. Do you see more convenient way to do it, or it doesn't look like a benefit to you?

**Comment by pugachAG** [comment]
> there is no way to distinguish between cases "delta was saved but it is empty" and "delta was not saved yet"

when do we need to distinguish between those cases?

**Comment by Longarithm** [comment]
> when do we need to distinguish between those cases?

We need to do it in `test_flat_storage_creation`. One of checked behaviours is that we do not save deltas before migration start, but do save deltas after that. Why this is needed is documented (or should be) in `FlatStorageCreationStatus`.

**Comment by pugachAG** [comment]
@Longarithm 

In this case I suggest creating a separate column for that to store flat state delta metadata.

**Comment by Longarithm** [comment]
It makes sense, then we can have two separate methods `FlatStateDelta::read_from_store` and `FlatStateDelta::exists` reading from these separate columns.

@mzhangmzz does it make sense to you - because you took part in implementing some of that logic?

**Comment by Longarithm** [comment]
We reconsidered approach to https://github.com/near/nearcore/issues/8436 (summary is updated there) and decided not to proceed with it. For MVP, deltas as a single DB entry are fine.

**Pull request closed without merging**